File: skills/code-architect/references/code-examples/ruby/visitor.rb
# frozen_string_literal: true

# Visitor — separate algorithms from objects via accept/visit (double dispatch).

# Elements expose #accept and call back the type-specific visitor method.
class Literal
  attr_reader :value

  def initialize(value) = @value = value
  def accept(visitor) = visitor.visit_literal(self)
end

class Addition
  attr_reader :left, :right

  def initialize(left, right)
    @left = left
    @right = right
  end

  def accept(visitor) = visitor.visit_addition(self)
end

# A visitor implements one method per element type.
class Evaluator
  def visit_literal(node) = node.value
  def visit_addition(node) = node.left.accept(self) + node.right.accept(self)
end

class Printer
  def visit_literal(node) = node.value.to_s
  def visit_addition(node) = "(#{node.left.accept(self)} + #{node.right.accept(self)})"
end

if __FILE__ == $PROGRAM_NAME
  # (1 + (2 + 3))
  tree = Addition.new(Literal.new(1), Addition.new(Literal.new(2), Literal.new(3)))

  puts "print: #{tree.accept(Printer.new)}"
  puts "eval:  #{tree.accept(Evaluator.new)}"
end

File: skills/code-architect/references/code-examples/python/visitor.py
"""Visitor — separate algorithms from objects via accept/visit (double dispatch)."""

from __future__ import annotations

from abc import ABC, abstractmethod


class Visitor(ABC):
    @abstractmethod
    def visit_text(self, element: "TextElement") -> str:
        ...

    @abstractmethod
    def visit_image(self, element: "ImageElement") -> str:
        ...


class Element(ABC):
    @abstractmethod
    def accept(self, visitor: Visitor) -> str:
        ...


class TextElement(Element):
    def __init__(self, content: str) -> None:
        self.content = content

    def accept(self, visitor: Visitor) -> str:
        return visitor.visit_text(self)


class ImageElement(Element):
    def __init__(self, src: str) -> None:
        self.src = src

    def accept(self, visitor: Visitor) -> str:
        return visitor.visit_image(self)


class HtmlExportVisitor(Visitor):
    def visit_text(self, element: TextElement) -> str:
        return f"<p>{element.content}</p>"

    def visit_image(self, element: ImageElement) -> str:
        return f'<img src="{element.src}"/>'


if __name__ == "__main__":
    elements: list[Element] = [TextElement("hi"), ImageElement("logo.png")]
    visitor = HtmlExportVisitor()
    for element in elements:
        print(element.accept(visitor))
